Sources/Commands/Workspace.swift

```diff
@@ -408,6 +414,68 @@ public class Workspace {
+    ///   - dependency: The dependency to pin.
+    ///   - packageName: The name of the package which is being pinned.
+    ///   - at: The version to pin at.
+    ///   - reason: The optional reason for pinning.
+    /// - Throws: WorkspaceOperationError, PinOperationError
+    func pin(dependency: ManagedDependency, packageName: String, at version: Version, reason: String? = nil) throws {
+        // Compute constaints with the new pin and try to resolve dependencies. We only commit the pin if the
+        // dependencies can be resolved with new constraints.
+        //
+        // The constraints consist of three things:
+        // * Root manifest
+        // * Exisiting pins except the dependency we're currently pinning.
+        // * The constraint for the new pin we're trying to add.
+        let constraints = computeRootPackageConstraints(try loadRootManifest()) 
+                        + pinsStore.createConstraints().filter({ $0.identifier != dependency.repository }) as [RepositoryPackageConstraint]
+                        // FIXME: This is broken, successor isn't correct and should be eliminated.
+                        + [RepositoryPackageConstraint(container: dependency.repository, versionRequirement: .range(version..<version.successor()))]
+        // Resolve the dependencies.
+        let results = try resolveDependencies(constraints: constraints)
+        // Add the record in pins store.
+        try pinsStore.pin(package: packageName, repository: dependency.repository, at: version, reason: reason)
+        // Update the checkouts based on new dependency resolution.
+        try updateCheckouts(with: results)
+    }
+
+    /// Pins all of the dependencies to the loaded version.
+    ///
+    /// - Parameters:
+    ///   - reason: The optional reason for pinning.
+    func pinAll(reason: String? = nil) throws {
+        // Load the package graph
+        _ = try loadPackageGraph()
+        // Load the dependencies.
+        let dependencyManifests = try loadDependencyManifests()
+        // Start pinning each dependency.
+        for dependencyManifest in dependencyManifests.dependencies {
+            let dependency = dependencyManifest.dependency
+            // Get package name.
+            let package = dependencyManifest.manifest.name
+            // If the dependency is in editable state, do not pin it.
+            guard !dependency.isInEditableState else {
+                print("warning: not pinning \(package) because it is being edited.")
+                continue
+            }
+            // We should have a version loaded to pin. This will never happen right now
+            // because we can't have dependencies checked out to a git ref.
+            guard let version = dependency.currentVersion else {
+                print("warning: not pinning \(package) because doesn't have a version loaded.")
+                continue
+            }
+            // Commit the pin.
+            try pinsStore.pin(
```

Comment:
loadPackageGraph() method call will resolve the current package graph (and fail if unresolvable) and then iterate over loaded dependencies and pin them directly in pinsStore (provided they can be pinned i.e have a version and not in editable state).